pkg/server/server.go

```diff
@@ -212,32 +220,65 @@ func (s *Server) Write(ctx context.Context, req *openfgapb.WriteRequest) (*openf
 }
 
 func (s *Server) Check(ctx context.Context, req *openfgapb.CheckRequest) (*openfgapb.CheckResponse, error) {
+
 	store := req.GetStoreId()
 	tk := req.GetTupleKey()
 	ctx, span := tracer.Start(ctx, "check", trace.WithAttributes(
-		attribute.KeyValue{Key: "store", Value: attribute.StringValue(req.GetStoreId())},
+		attribute.KeyValue{Key: "store", Value: attribute.StringValue(store)},
 		attribute.KeyValue{Key: "object", Value: attribute.StringValue(tk.GetObject())},
 		attribute.KeyValue{Key: "relation", Value: attribute.StringValue(tk.GetRelation())},
 		attribute.KeyValue{Key: "user", Value: attribute.StringValue(tk.GetUser())},
 	))
 	defer span.End()
 
+	if tk.GetUser() == "" || tk.GetRelation() == "" || tk.GetObject() == "" {
+		return nil, serverErrors.InvalidCheckInput
+	}
+
 	modelID, err := s.resolveAuthorizationModelID(ctx, store, req.GetAuthorizationModelId())
 	if err != nil {
 		return nil, err
 	}
 
-	q := commands.NewCheckQuery(s.datastore, s.meter, s.logger, s.config.ResolveNodeLimit)
+	model, err := s.datastore.ReadAuthorizationModel(ctx, store, modelID)
+	if err != nil {
+		return nil, err
+	}
 
-	res, err := q.Execute(ctx, &openfgapb.CheckRequest{
-		StoreId:              store,
-		TupleKey:             tk,
-		ContextualTuples:     req.GetContextualTuples(),
-		AuthorizationModelId: modelID,
-		Trace:                req.GetTrace(),
+	typesys := typesystem.New(model)
+
+	if err := validation.ValidateUserObjectRelation(typesys, tk); err != nil {
+		return nil, serverErrors.ValidationError(err)
+	}
+
+	for _, ctxTuple := range req.GetContextualTuples().GetTupleKeys() {
+		if err := validation.ValidateTuple(typesys, ctxTuple); err != nil {
+			return nil, serverErrors.HandleTupleValidateError(err)
+		}
+	}
+
+	ctx = typesystem.ContextWithTypesystem(ctx, typesys)
+	ctx = storage.ContextWithContextualTuples(ctx, req.ContextualTuples.GetTupleKeys())
+
+	resp, err := s.checkResolver.ResolveCheck(ctx, &graph.ResolveCheckRequest{
+		StoreID:              req.GetStoreId(),
+		AuthorizationModelID: req.GetAuthorizationModelId(),
+		TupleKey:             req.GetTupleKey(),
+		ContextualTuples:     req.ContextualTuples.GetTupleKeys(),
+		ResolutionMetadata: &graph.ResolutionMetadata{
+			Depth: s.config.ResolveNodeLimit,
+		},
 	})
 	if err != nil {
-		return nil, err
+		if errors.Is(err, graph.ErrResolutionDepthExceeded) {
+			return nil, serverErrors.AuthorizationModelResolutionTooComplex
+		}
+
+		return nil, serverErrors.HandleError("", err)
+	}
+
+	res := &openfgapb.CheckResponse{
+		Allowed: resp.Allowed,
```

Comment:
Should we add a span attribute here with `allowed: resp.Allowed`?

Reply:
Sure!

Reply:
👍